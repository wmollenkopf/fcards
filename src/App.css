Body {
  text-align: center;
  background:
    radial-gradient(black 15%, transparent 16%) 0 0,
    radial-gradient(black 15%, transparent 16%) 8px 8px,
    radial-gradient(rgba(255, 255, 255, .1) 15%, transparent 20%) 0 1px,
    radial-gradient(rgba(255, 255, 255, .1) 15%, transparent 20%) 8px 9px;
  background-color: #282828;
  background-size: 16px 16px;
  color: #DDD;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}