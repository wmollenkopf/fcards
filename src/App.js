import React, { Component } from 'react';
import './App.css';
import Signin from './components/Signin/Signin';
import Register from './components/Register/Register';
import CardOptions from './components/CardOptions/CardOptions';
import CardQuestion from './components/Card/CardQuestion/CardQuestion';
import CardAnswer from './components/Card/CardAnswer/CardAnswer';
import ShowAnswerButton from './components/ShowAnswerButton/ShowAnswerButton';
import Navigation from './components/Navigation/Navigation';
import { QUESTION_CARD, ANSWER_CARD } from './constants/CardTypes'


const initialState = {
  sessionKey:'boop',
  allCards: [],
  card: {
    card_id: 0,
    card_question: '',
    card_answer: ''
  },
  currentCardIndex: 0,
  showAnswer: false,
  route: 'signin',
  isSignedIn: false,
  user: {
    user_id: 4,
    username: 'biri@biri.me',
    joined: ''
  },
  editing: false
}

class App extends Component {
  constructor() {
    super();
    this.state = initialState;
  };

  loadUser = (data) => {
    this.setState({user: {
      id: data.user_id,
      username: data.username,
      joined: data.created_at
    }})
}

onRouteChange = (route) => {
  if (route === 'signout') {
    this.setState(initialState)
  } else if (route === 'home') {
    this.setState({isSignedIn: true})
  } else {
    this.setState({route: route});
  }
  
}


componentDidMount() {
  // Go ahead and load all of the cards for this user, if the user is logged in...
  if(this.state.user) {
    this.getCards();
  }
}

getCards = () => {
  console.log("Start");
  fetch('http://localhost:3000/getcards', {
    method: 'post',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({
      userId: this.state.user.user_id,
      sessionKey: this.state.sessionKey
    })
  })
    .then(response => response.json())
    .then(data => {
      console.log("data",data);
      this.setState({allCards: data});
      if(data.length>0){
        this.setState({card: data[0]});
        this.setState({currentCardIndex: 0});
      }
      
      return data;
    })
    .catch(err => console.log(err))
}

prevCard = () => {
  
  let currentIndex = this.state.currentCardIndex;
  if((currentIndex-1)<= 0) {
    currentIndex = (this.state.allCards.length-1);  
    this.setState({currentCardIndex: currentIndex});
  } else {
    currentIndex--;
  }
  
  this.setState({card: this.state.allCards[currentIndex]});
  this.setState({currentCardIndex: (currentIndex)});
  this.setState({showAnswer: false});
  console.log(this.state.card);
  console.log("Current Index Now1:",this.state.currentCardIndex);
}

nextCard = () => {

  let currentIndex = this.state.currentCardIndex;
  
  if((currentIndex+1)>=this.state.allCards.length) {
      this.setState({currentCardIndex: 0});
      currentIndex = 0;
  } else {
    currentIndex++;
  }
  
  this.setState({card: this.state.allCards[currentIndex]});
  this.setState({currentCardIndex: (currentIndex)});
  this.setState({showAnswer: false});

  console.log(this.state.card);
  console.log("Current Index Now2:",this.state.currentCardIndex);
}

showAnswerCard = () => {
  console.log("Showing answer...");
  this.setState({showAnswer: true});
}

showQuestionCard = () => {
  console.log("Showing question...");
  this.setState({showAnswer: false});
}

enableEditing = () => {
  console.log("Enabling Editing...");
  this.setState({editing: true});
}

saveEditing = () => {
  console.log(`Saving...`);
  console.log(this.state.card);
  
  // TODO
  this.setState({editing: false});
  fetch('http://localhost:3000/editcard', {
    method: 'post',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(
      {
        sessionKey: this.state.sessionKey,
        "cardId": this.state.card.card_id,
        "cardQuestion": this.state.card.card_question,
        "cardAnswer":this.state.card.card_answer
      }
      )
  })
    .then(response => response.json())
    .then(data => {
      console.log("new data",data);
      //return data;
    })
    //.then(this.getCards()) // might not need this for editing actually OR adding cards actually, maybe deleting?
    .then(() => {
        const theDeck = this.state.allCards;
        theDeck[this.state.currentCardIndex] = this.state.card;
        return theDeck;
    })
      .then((theDeck)=>{
        this.setState({allCards: theDeck});
      })
      .then(console.log('statedeck:',this.state.allCards))
    .catch(err => console.log(err))
}


handleEditing = (type) => (event) => {
  if(type===QUESTION_CARD) {
    const newCard = Object.assign({},this.state.card,{card_question: event.target.value});
    this.setState({card: newCard});
  } else if(type===ANSWER_CARD) {
    const newCard = Object.assign({},this.state.card,{card_answer: event.target.value});
    this.setState({card: newCard});
  }
  
}

//<ShowAnswerButton showAnswerCard={this.showAnswerCard} showAnswer={showAnswer} />
  render() {
    console.log("Current Index Now3:",this.state.currentCardIndex);
    const { isSignedIn, card, route,showAnswer,editing } = this.state;
    return (
      <div className="App">
        <Navigation isSignedIn={isSignedIn} onRouteChange={this.onRouteChange} />   
        { route !== 'home'
          ? (
            this.state.allCards.length > 0
            ? 
            <div >
              <CardOptions prevCard={this.prevCard} nextCard={this.nextCard} enableEditing={this.enableEditing} />
              <CardQuestion card={card} showAnswer={showAnswer} editing={editing} saveEditing={this.saveEditing} handleEditing={this.handleEditing} />
              <CardAnswer card={card} showAnswer={showAnswer} editing={editing} saveEditing={this.saveEditing} handleEditing={this.handleEditing}/>
              <ShowAnswerButton showAnswerCard={this.showAnswerCard} showAnswer={showAnswer} nextCard={this.nextCard} />
            </div>
            :<div>No Cards Loaded Yet</div>            
            )
          : (
             route === 'signin'
             ? <Signin loadUser={this.loadUser} onRouteChange={this.onRouteChange}/>
             : <Register loadUser={this.loadUser} onRouteChange={this.onRouteChange}/>
            )
          }     
        
      </div>
    );
  }
}

export default App;
