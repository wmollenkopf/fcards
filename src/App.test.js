import React from 'react';
import {shallow} from 'enzyme';
import App from './App';

const initialCard = {
  card_id: 0,
  card_question: '',
  card_answer: ''
};

const initialState = {
  sessionKey:'',
  allCards: [],
  card: initialCard,
  currentCardIndex: 0,
  showAnswer: false,
  route: 'signin',
  isSignedIn: false,
  editing: false,
  creating: false,
  fcardServerURL: `https://fcards-server.biri.me`,
}

const app = shallow(<App/>);
describe('App',()=>{

    it('renders correctly', ()=>{
        expect(app).toMatchSnapshot();
    });
    
    describe('When state is first initialized',()=>{});
    it('initializes the `state` `fcardServerURL` with a URL',()=>{
        expect(app.state().fcardServerURL);
    });
 

    

})


// it('renders without crashing', () => {
//   const div = document.createElement('div');
//   ReactDOM.render(<App />, div);
//   ReactDOM.unmountComponentAtNode(div);
// });


