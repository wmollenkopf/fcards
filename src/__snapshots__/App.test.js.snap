// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Navigation
          isSignedIn={false}
          performSignOut={[Function]}
        />,
        <Signin
          allCards={Array []}
          enableCreating={[Function]}
          fcardServerURL="http://localhost:3000"
          loadUser={[Function]}
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "isSignedIn": false,
          "performSignOut": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allCards": Array [],
          "enableCreating": [Function],
          "fcardServerURL": "http://localhost:3000",
          "loadUser": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Navigation
            isSignedIn={false}
            performSignOut={[Function]}
          />,
          <Signin
            allCards={Array []}
            enableCreating={[Function]}
            fcardServerURL="http://localhost:3000"
            loadUser={[Function]}
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "isSignedIn": false,
            "performSignOut": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allCards": Array [],
            "enableCreating": [Function],
            "fcardServerURL": "http://localhost:3000",
            "loadUser": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
